fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -46,6 +45,21 @@ class ShardedDataParallel(nn.Module):
             Synchronize the models in between the ranks when starting up. Not needed if each rank has the same seed,
             or the training restarts from a saved state
 
+
+    .. warning:
+        ShardedDDP implements gradient sharding, meaning that each rank only owns a unique shard of the model gradients
+        after the backward pass, in order to save memory and some communication bandwidth.
+
+    .. warning:
+        As a consequence of sharding, in case of gradient clipping, one has to use the `clip_grad_norm` exposed by
```

Comment:
If users miss this docstring, will they hit an error message? How can we communicate this and the below hanging behavior to users? my fear is that they may miss these warnings and end up with a hanging run which are hard to debug.

Reply:
it's hard to catch because this is typically done outside of OSS/ShardedDDP, what happens is that the grad scaler will check the gradients (on each rank), and will bail out if there are NaNs/inf. If it's not the case on all ranks, then some of them bail out of the step and others repeat the step, they're not in sync and the communication primitives (broadcast) hang and time out. 
I'm not sure how to improve on that, some frameworks (lightning and huggingface for instance) help with their fairscale integration because they handle all of that and the user does not have to know

Reply:
Thanks for the context! We will have to instrument timeouts on collective operations to handle hanging issues such as these. Definitely tricky to handle all the cases though since we don't have visibility.